### ISSUE TITLE
[Go][Arrow] Remove `panic` from `arrow.Schema` and `arrow.Metadata` construction

### ISSUE THREAD

**Issue by ggodik** [apache/arrow#37590]
### Describe the enhancement requested

Prefer to return an `error` instead of `panic` as this situation is due to a bad user input.
Change the signature to return a `Schema`/`Metadata` and an `error`

list of `panic`s

- https://github.com/apache/arrow/blob/main/go/arrow/schema.go#L182
- https://github.com/apache/arrow/blob/main/go/arrow/schema.go#L34

### Component(s)

Go

**Comment by ggodik** [comment]
I'm working on a PR

**Comment by ggodik** [comment]
@zeroshade 

there's a lot of places where `arrow.NewSchema` is inlined in unit tests

what do you think about a helper to wrap it 

```
// NoError is a helper function that asserts that the error is nil and returns the value.
func NoError[S any](t *testing.T) func(s S, err error) S {
	return func(s S, err error) S {
		assert.NoError(t, err)
		return s
	}
}
```

**Comment by zeroshade** [comment]
I like it. Besides i think it's time we start leveraging generics in the arrow lib and bump to go1.18 anyways :smile:

**Comment by ggodik** [comment]
The `mod` file is already at 1.20 https://github.com/apache/arrow/blob/main/go/go.mod#L19C1-L19C8

**Comment by zeroshade** [comment]
@ggodik yes, but we've been maintaining compatibility with go1.17 still via build tags. Sections like the compute lib which used generics have had build tags to ensure that they would simply be excluded in go1.17 builds and we have CI still building that version.

So i'd be fine with removing the go1.17 CI builds and officially dropping support for it now.
